### PR TITLE
Improve the supervisor app state GET v3 tests

### PR DESCRIPTION
**Pull request by thgreasi**
Change-type: patch

### CODE REVIEW

File: test/19_apps.ts [hunk 1]
```diff
@@ -20,23 +23,31 @@ describe('Apps', function () {
...
+			supervisorApp = fx.applications.supervisorApp;
+			supervisorRelease2 = fx.releases.supervisorRelease2;
 
 			deviceWithSupervisor = await fakeDevice.provisionDevice(
 				admin,
-				applicationId,
+				userApp.id,
 				'balenaOS 2.3.0+rev1',
 				'1.0.1',
-				fx.releases.supervisorRelease1.id,
```
Comment (thgreasi) [inline test/19_apps.ts:32]: Ends up that all that this was not respected and the API was overwriting it w/ the sv release that it found based on `1.0.1`. On top of this `provisionDevice` should probably be changed to be using `register/device` rather than orchestrating OData requests, so that it's more realistic, and in that case providing the target sv release will not be a case anyway.

File: test/test-lib/api-helpers.ts [hunk 2]
```diff
@@ -81,33 +82,45 @@ export const addImageToRelease = async (
...
+
+	const pineClientRequest = pineUser.get({
+		resource,
+		id,
+		options: {
+			$select: Object.keys(expectations),
+		},
+	});
+	const result = (
+		'expect' in pineClientRequest
```
Comment (fisehara) [inline test/test-lib/api-helpers.ts:110]: Uncommon, but do I understand correctly that the pineClientRequest has a key `expect` ?

File: test/test-lib/api-helpers.ts [hunk 3]
```diff
@@ -81,33 +82,45 @@ export const addImageToRelease = async (
...
+			? (await (pineClientRequest as ReturnType<PineTest['get']>).expect(200))
+					.body
+			: await pineClientRequest
+	) as T | undefined;
+	expect(result).to.be.an('object');
+	Object.entries(expectations).forEach(([key, valueOrAssetion]) => {
```
Comment (fisehara) [inline test/test-lib/api-helpers.ts:116]: Why is that changed from the `for ([] of ) {}` ?

File: test/test-lib/api-helpers.ts [hunk 4]
```diff
@@ -81,33 +82,45 @@ export const addImageToRelease = async (
...
+		if (typeof valueOrAssetion === 'function') {
+			valueOrAssetion(expect(result).to.have.property(key));
+		} else if (typeof valueOrAssetion === 'object' && valueOrAssetion != null) {
+			expect(result).to.have.property(key).to.deep.equal(valueOrAssetion);
 		} else {
-			expect(result).to.have.property(key, valueOrAssertion);
+			expect(result).to.have.property(key, valueOrAssetion);
 		}
```
Comment (fisehara) [inline test/test-lib/api-helpers.ts:123]:
```suggestion
		if (typeof valueOrAssertion === 'function') {
			valueOrAssertion(expect(result).to.have.property(key));
		} else if (typeof valueOrAssertion === 'object' && valueOrAssertion != null) {
			expect(result).to.have.property(key).to.deep.equal(valueOrAssertion);
		} else {
			expect(result).to.have.property(key, valueOrAssertion);
		}
```

File: test/test-lib/api-helpers.ts (hunk 2 above)
Comment (thgreasi) [reply test/test-lib/api-helpers.ts:110]: So at this point pineUser is either a pinejs-client-supertest or a pinejs-client-request (the one the pinejs exposes) instance, and the badly named `pineClientRequest` variable here is a request object (promise). So if it's from `pinejs-client-supertest`, then it will also have a `.expect()` (since it's a supertest + pinejs-client hybrid). So here we are doing a feature check (rather than instance check) and if `.expect` is available, we use it. (let me rename the var...)

File: test/test-lib/api-helpers.ts (hunk 3 above)
Comment (thgreasi) [reply test/test-lib/api-helpers.ts:116]: copied over from bC-api, will make it a for-of again :+1:

**Review by fisehara — APPROVED** [review]

File: test/test-lib/api-helpers.ts [hunk 5]
```diff
@@ -1,8 +1,9 @@
+import type { PinejsClient } from '@balena/pinejs/out/sbvr-api/sbvr-utils';
```
Comment (Page-) [inline test/test-lib/api-helpers.ts:1]: @thgreasi can you remove the deep imports like this please

Comment (thgreasi) [reply test/test-lib/api-helpers.ts:1]: OOOoppps, sorry for that.
See: https://github.com/balena-io/open-balena-api/pull/1500
